### PR TITLE
Annotate channels with some ordered metadata

### LINKED ISSUE #12880: Annotate ordered channel metadata during content annotation
## Current implementation

Currently Kolibri annotates a few different top level metadata fields onto channels after successful import: https://github.com/learningequality/kolibri/blob/develop/kolibri/core/content/utils/annotation.py#L770

Specifically, this function sets the languages of contentnodes in the included languages of the channel: https://github.com/learningequality/kolibri/blob/develop/kolibri/core/content/utils/annotation.py#L818

This is done using a Django many to many Field (M2MField).

## Desired changes

For 0.18, we would like to show additional metadata on channel cards - specifically the categories and grade_levels metadata, aggregated across all the resources in the channel.

To achieve this we should add categories and grade_levels comma separated TextFields to the ChannelMetadata model. We will use the TextField to maintain a parallel implementation to the ContentNode model.

However, due to the fact that this could result in an overwhelming amount of metadata, we will be prioritizing the display of this metadata - limiting to the top 3 categories, top 3 grade_levels, and the most common language. This means that our annotation needs to reflect this fact.

To achieve this for the TextFields, we will simply order the categories and grade_levels in the comma separated text_fields by their relative resource counts (with the category or grade_level with the highest number of resources going first in its respective field, descending from there).

To achieve this for the included_languages field, we will need a sorted M2M field implementation. We can either roll our own, or use this package: https://pypi.org/project/django-sortedm2m/3.1.1/ (this is the last version to support Django 3.2) to update the field to a sorted field.

If using the above package, we should carefully follow their instructions for creating the database migration.

Finally, an upgrade function should be implemented using Kolibri's upgrade mechanisms for any old_version less than "0.18.0" to rerun the annotation for every channel. This should be put into the `upgrade.py` of the core content app, which also contains an exuberance of examples of these kinds of functions: https://github.com/learningequality/kolibri/blob/develop/kolibri/core/content/upgrade.py

### PR DESCRIPTION
**Pull request by jredrejo**
## Summary
- In `ChannelMetadata`, it creates two new fields:
  - `included_categories`
  - `included_grade_levels`
  - and changes type of `included_languages` from ManyToManyField to a special type (from a new added library dependency) called `SortedManyToManyField`
- Creates migrations for these fields
- Creates and/or modify `ChannelMetadata` annotations 
  - `calculate_included_languages`
  - `calculate_ordered_categories`
  - `calculate_ordered_grade_levels`
- Adds tests for these changes
- Adds an upgrade method, `ordered_metadata_in_channels`,  for previously annotated channels  (Important note: this has to be skipped in KDP migrations)

## References
Fixes: #12880

## Reviewer guidance
- do tests pass
- try to import channels, totally or partially and check their included_categories, included_grade_levels and included_languages are updated correctly

### CODE REVIEW

File: kolibri/core/content/utils/annotation.py [hunk 1]
```diff
@@ -810,11 +813,50 @@ def calculate_total_resource_count(channel):
...
+        channel_id=channel.id, available=True
+    ).exclude(**{field_name: None})
+    all_values = []
+    for node in content_nodes.values_list(field_name, flat=True):
+        if node:  # just in case some field is an empty string
+            all_values.extend(node.split(","))
+
+    value_counts = {}
+    for value in all_values:
+        value_counts[value] = value_counts.get(value, 0) + 1
```
Comment (rtibbles) [inline kolibri/core/content/utils/annotation.py:827]: This is a beautifully concise way to default the value to 0.

**Review by rtibbles — COMMENTED** [review]
If you can refer to the big comment at the top of models.py - we want to run this command:
```
kolibri manage generate_schema current
```
and commit the result as part of this PR too.

**Comment by rtibbles** [comment]
I thought I had fixed this last remaining flaky content test failure on postgres, but it still happened! Have rerun it.

**Comment by rtibbles** [comment]
Looking good to me! I'll give it a bit more of a close review on Monday. If you can do the schema regeneration, I think this should be quickly mergeable.

**Comment by jredrejo** [comment]
> If you can refer to the big comment at the top of models.py - we want to run this command:
> 
> ```
> kolibri manage generate_schema current
> ```

done, but had to modify a bit `generate_schema` due to python 3.12 compatibility issues

File: kolibri/core/content/management/commands/generate_schema.py [hunk 2]
```diff
@@ -10,6 +11,51 @@
...
+            args=spec.args,
+            varargs=spec.varargs,
+            keywords=spec.varkw,
+            defaults=spec.defaults,
+        )
+
+    inspect.ArgSpec = ArgSpec
+    inspect.getargspec = getargspec
+
+# SQLAlchemy 2.0 compatibility patch
```
Comment (rtibbles) [inline kolibri/core/content/management/commands/generate_schema.py]: Why do we need a compatibility patch with SQLAlchemy 2? We explicitly pin to less than 2, and I've not had any issues installing SQLAlchemy 1.4.49 in Python 3.12.

File: kolibri/core/content/contentschema/versions/content_schema_current.py [hunk 3]
```diff
@@ -45,6 +46,9 @@ class ContentLocalfile(Base):
 class ContentContentnode(Base):
     __tablename__ = "content_contentnode"
     __table_args__ = (
+        CheckConstraint(
```
Comment (rtibbles) [inline kolibri/core/content/contentschema/versions/content_schema_current.py]: Adding the constraint checks seems to be causing the Python test errors.

File: kolibri/core/content/management/commands/generate_schema.py (hunk 2 above)
Comment (jredrejo) [reply kolibri/core/content/management/commands/generate_schema.py]: fixed, somehow I mixed my virtual environments

File: kolibri/core/content/contentschema/versions/content_schema_current.py (hunk 3 above)
Comment (jredrejo) [reply kolibri/core/content/contentschema/versions/content_schema_current.py]: Fixed

**Review by rtibbles — APPROVED** [review]
Everything is to spec - merging so we can use this as the basis for other work as well!

**Pull request merged**